File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode* root, int curr, int &sum) {
        if(!root) return;
        curr = curr*10 + root->val;
        if(!root->left && !root->right) {
            sum += curr;
        }

        solve(root->left, curr, sum);
        solve(root->right, curr, sum);
    }
    
    int sumNumbers(TreeNode* root) {
        if(!root) return 0;
        int sum = 0;
        solve(root, 0, sum);
        return sum;
    }
};

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void solve(TreeNode root, int curr, int[] sum) {
        if(root == null) return;
        curr = curr * 10 + root.val;
        if(root.left == null && root.right == null) {
            sum[0] += curr;
        }

        solve(root.left, curr, sum);
        solve(root.right, curr, sum);
    }

    public int sumNumbers(TreeNode root) {
        if(root == null) return 0;
        int[] sum = new int[1];
        solve(root, 0, sum);
        return sum[0];
    }
}